src/cdk/a11y/focus-trap/focus-trap-manager.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable} from '@angular/core';
+
+/**
+ * A FocusTrap managed by FocusTrapManager.
+ * Implemented by ConfigurableFocusTrap to avoid circular dependency.
+ */
+export interface ManagedFocusTrap {
+    enable(): void;
+    disable(): void;
+    focusInitialElementWhenReady(): void;
+}
+
+/** Injectable that ensures only the most recently enabled FocusTrap is active. */
+@Injectable({providedIn: 'root'})
+export class FocusTrapManager {
+  // A stack of the FocusTraps on the page. Only the FocusTrap at the
+  // top of the stack is active.
+  private _focusTrapStack: ManagedFocusTrap[] = [];
+
+  /**
+   * Disables the FocusTrap at the top of the stack, and then pushes
+   * the new FocusTrap onto the stack.
+   */
+  register(focusTrap: ManagedFocusTrap): void {
+    // Dedupe focusTraps that register multiple times.
```

Comment:
Under what circumstances could a trap register itself more than once?

Reply:
If a developer sets enabled=true multiple times, or creates a ConfigurableFocusTrap (since it defaults to enabled=true, it registers itself in the constructor) and then sets enabled=true anyways.